### ISSUE TITLE
Transparent faces in mesh plots overlap incorrectly

### ISSUE THREAD

**Issue by scottsleep** [plotly/plotly.js#3243]
Normally, when a face is opaque it can obscure other faces, or if the camera is rotated 180 degrees the first is now obscured by the second. This is currently not true for faces with opacity < 1. Instead it seems as if they are always layered in order of creation, so the first face in the mesh will always be on top (or beneath?) the others, no matter their relative 3D location.

The example contains 4 triangles: the two on the left are opaque while the two on the right are transparent.

- Notice how the orange face is always in front of the blue one, even though they intersect and part of it should be behind, as demonstrated by the purple and red triangles.
- Even if the camera is rotated.

![plotly bug](https://user-images.githubusercontent.com/42533046/48465563-2239f580-e833-11e8-9369-fbad3b2eaaf7.PNG)
https://codepen.io/anon/pen/bQqeMg

This leads to situations where obscured objects are drawn in front. All the boxes in the image below are either behind or inside the Blue one.
![plotly bug2](https://user-images.githubusercontent.com/42533046/48465756-ce7bdc00-e833-11e8-990b-f08bcb9029fa.PNG)

**Comment by etpinard** [comment]
Yeah, that's essentially the `mesh3d` manifestation of bug https://github.com/plotly/plotly.js/issues/1267

Thank you for writing in with such clear examples.

**Comment by etpinard** [comment]
Related https://github.com/plotly/plotly.js/issues/3138

**Comment by etpinard** [comment]
Another example: https://github.com/plotly/plotly.js/issues/3410 - this for `surface` traces.

**Comment by archmoj** [comment]
I think we may need to sort triangles when there are transparent ones. There might be other solutions useful for orthographic projection {[this PR](https://github.com/plotly/plotly.js/pull/3331)}. 
Another idea for making surfaces transparent is surface tessellation.

**Comment by archmoj** [comment]
Another [codepen](https://codepen.io/MojtabaSamimi/pen/gqQazW?editors=0010) to highlight the draw order issue still exists even with lower opacity values and separated RGB channels.

**Comment by archmoj** [comment]
OK. I think a patch like [this commit](https://github.com/gl-vis/gl-plot3d/commit/3e0b73e99b0ef2120eb765dbb23759fd9114710b) may help. 
cc: https://github.com/plotly/plotly.js/issues/4111
cc: https://github.com/plotly/plotly.js/issues/1267

Please notice how one could get similar overlap colors [here](https://codepen.io/MojtabaSamimi/pen/KOQaBN?editors=0010) but not [before](https://codepen.io/MojtabaSamimi/pen/gqQazW?editors=0010).

Here are more examples:
example 2
[before](https://codepen.io/MojtabaSamimi/pen/OKQpzW?editors=0010)
[after](https://codepen.io/MojtabaSamimi/pen/jgZwbE?editors=0010)

example 3
[before](https://codepen.io/MojtabaSamimi/pen/dxdWLZ)
[after](https://codepen.io/MojtabaSamimi/pen/MNQmLY)

example 4
[before](https://codepen.io/MojtabaSamimi/pen/KOQmOK?editors=0010)
[after](https://codepen.io/MojtabaSamimi/pen/GVQrML?editors=0010)

**Comment by archmoj** [comment]
Also for `scatter3d` traces there is another possibility to enable depth like these:
[example 2](https://codepen.io/MojtabaSamimi/pen/PMRqjO)
[example 3](https://codepen.io/MojtabaSamimi/pen/GVxJEv)
[example 4](https://codepen.io/MojtabaSamimi/pen/KOopqr)

**Comment by archmoj** [comment]
And finally for regular point-clouds [here](https://github.com/gl-vis/gl-plot3d/commit/6bb28531e2b2a5c290cc252d675d9b3e7b4a910b) is the algorithm that works best:
[Example 1](https://codepen.io/MojtabaSamimi/pen/pMLger)
[Example 2](https://codepen.io/MojtabaSamimi/pen/zgWrNa)
[Example 3](https://codepen.io/MojtabaSamimi/pen/EqEPZp)
[Example 4](https://codepen.io/MojtabaSamimi/pen/JgLGEe)

cc: #4111
cc: #1267

**Comment by archmoj** [comment]
Also for `surface` plot the app at https://dash-gallery.plotly.host/dash-yield-curve/ displays  
![Screenshot from 2019-11-11 10-07-42](https://user-images.githubusercontent.com/33888540/68597658-5a3a3f80-046b-11ea-8906-191e27251779.png)

**Comment by jackparmer** [comment]
This issue has been tagged with `NEEDS SPON$OR`

A [community PR](https://github.com/plotly/plotly.js/pulls?q=is%3Apr+is%3Aopen+label%3A%22type%3A+community%22) for this feature would certainly be welcome, but our experience is deeper features like this are difficult to complete without the Plotly maintainers leading the effort.

**Sponsorship range: $15k-$20k**

What Sponsorship includes:
- Completion of this feature to the Sponsor's satisfaction, in a manner coherent with the rest of the Plotly.js library and API
- Tests for this feature
- Long-term support (continued support of this feature in the latest version of Plotly.js)
- Documentation at [plotly.com/javascript](https://plotly.com/javascript/)
- Possibility of integrating this feature with Plotly Graphing Libraries (Python, R, F#, Julia, MATLAB, etc)
- Possibility of integrating this feature with Dash
- Feature announcement on [community.plotly.com](https://community.plotly.com/tag/announcements) with shout out to Sponsor (or can remain anonymous)
- Gratification of advancing the world's most downloaded, interactive scientific graphing libraries (>50M downloads across supported languages)

Please include the link to this issue when [contacting us](https://plotly.com/get-pricing/) to discuss.

**Comment by gvwilson** [comment]
@archmoj is this currently on your pile? No worries if not - we're just trying to do some planning.

**Comment by archmoj** [comment]
@gvwilson yes it is. I need 2 days to possibly complete #4643 pull.
